build/commands/lib/util.js

```diff
@@ -606,48 +607,98 @@ const util = {
+
+    const doesBuildNinjaExist = fs.existsSync(path.join(outputDir, 'build.ninja'))
+    const hasBuildArgsUpdated = util.writeGnBuildArgs(outputDir, buildArgs)
 
-    const shouldRunGnGen = config.force_gn_gen ||
-      !fs.existsSync(buildNinjaFile) || !fs.existsSync(gnArgsFile) ||
-      !prevBuildArgs || prevBuildArgs != buildArgsStr ||
-      !prevExtraGnGenOpts || prevExtraGnGenOpts != config.extraGnGenOpts
+    const shouldRunGnGen =
+      config.force_gn_gen ||
+      !doesBuildNinjaExist ||
+      hasBuildArgsUpdated ||
+      gnGenGuard.isDirty()
 
     if (shouldRunGnGen) {
-      // `gn gen` can modify args.gn even if it's failed.
-      // Therefore delete the file to make sure that args.gn and
-      // brave_build_args.txt are in sync.
-      if (prevBuildArgs)
-        fs.removeSync(buildArgsFile)
+      gnGenGuard.run(() => {
+        util.run('gn', ['gen', outputDir, ...extraGnGenOpts], options)
+      })
+    }
+  },
+
+  writeGnBuildArgs: (outputDir, buildArgs) => {
+    // Generate build arguments in .gni format to be imported into args.gn. This
+    // approach enables customization of args.gn without the build scripts
+    // resetting it during each execution.
+    const generatedArgsContent = [
+      '# Do not edit, any changes will be lost on next build.',
+      '# To customize build args, use args.gn in the same directory.\n',
+      ...Object.entries(buildArgs).map(([arg, val]) => {
+        assert(typeof arg === 'string')
+        assert(val !== undefined)
+        if (val === null) {
+          // Output only arg, it may be a comment or an import statement.
+          return arg
+        }
+        return `${arg}=${JSON.stringify(val)}`
+      })
+    ].join('\n')
+
+    // Write the generated arguments to the args_generated.gni file. The file
+    // name is intentionally chosen to be close to args.gn.
+    const generatedArgsFilePath = path.join(outputDir, 'args_generated.gni')
+    const hasGeneratedArgsUpdated =
+      util.writeFileIfModified(generatedArgsFilePath, generatedArgsContent + '\n')
+    if (hasGeneratedArgsUpdated) {
+      Log.status(`${generatedArgsFilePath} has been updated`)
+    }
+
+    // Import args_generated.gni into args.gn.
+    const argsGnFilePath = path.join(outputDir, 'args.gn')
+    const generatedArgsImportLine =
+      `import("//${path.relative(config.srcDir, generatedArgsFilePath).replace(/\\/g, '/')}")`
+
+    let isArgsGnValid = false
+    if (fs.existsSync(argsGnFilePath)) {
+      const argsGnContent = fs.readFileSync(argsGnFilePath, {encoding: 'utf-8'})
+      // Use includes() method to ensure that the import statement from
+      // args_generated.gni is present in args.gn (even if the user has made
+      // modifications). This import statement can also be commented out,
+      // allowing the user to fully ignore generated arguments.
+      isArgsGnValid = argsGnContent.includes(generatedArgsImportLine)
```

Comment:
There is a case when a user overrides an arg listed in `args_generated.gni`.
If they add before `import` it will have no effect.
Let's check that the header + import are the first lines?

Reply:
Added a comment about the extra args placement instead (discussed in Slack).